_posts/2021-09-13-user-defined-value-types.md

```diff
@@ -0,0 +1,183 @@
+type with 18 decimals and a minimal library to do arithmetic operations on the type.
+
+```solidity
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.8;
+
+// Represent a 18 decimal, 256 bit wide fixed point type using a user defined value type.
+type UFixed256x18 is uint256;
+
+/// A minimal library to do fixed point operations on UFixed256x18.
+library FixedMath {
+    uint constant multiplier = 10**18;
+
+    /// Adds two UFixed256x18 numbers. Reverts on overflow, relying on checked
+    /// arithmetic on uint256.
+    function add(UFixed256x18 a, UFixed256x18 b) internal pure returns (UFixed256x18) {
+        return UFixed256x18.wrap(UFixed256x18.unwrap(a) + UFixed256x18.unwrap(b));
+    }
+    /// Multiplies UFixed256x18 and uint256. Reverts on overflow, relying on checked
+    /// arithmetic on uint256.
+    function mul(UFixed256x18 a, uint256 b) internal pure returns (UFixed256x18) {
+        return UFixed256x18.wrap(UFixed256x18.unwrap(a) * b);
+    }
+    /// Take the floor of a UFixed256x18 number.
+    /// @return the largest integer that does not exceed `a`.
+    function floor(UFixed256x18 a) internal pure returns (uint256) {
+        return UFixed256x18.unwrap(a) / multiplier;
+    }
+    /// Turns a uint256 into a UFixed256x18 of the same value.
+    /// Reverts if the integer is too large.
+    function toUFixed256x18(uint256 a) internal pure returns (UFixed256x18) {
+        return UFixed256x18.wrap(a * multiplier);
+    }
+}
+```
+
+Notice how `UFixed256x18.wrap` and `FixedMath.toUFixed256x18` have the same signature but perform
+two very different operations: The `UFixed256x18.wrap` function returns a `UFixed256x18` that has
+the same data representation as the input, whereas `toUFixed256x18` returns a `UFixed256x18` that
+has the same numerical value. One can allow some form of type-encapsulation by only using the `wrap`
+and `unwrap` functions in the file that defines the type.
+
+## Operators and Type Rules
+
+Explicit and implicit conversions to and from other types are disallowed.
+
+Currently, no operators are defined for user defined value types. In particular, even the operator
+`==` is not defined. However, allowing operators is currently being discussed. To give a short
+
+outlook on the applications, one may want to introduce a new integer type that always does wrapping
+arithmetic as follows:
+
+```solidity
+/// Proposal on defining operators on user defined value types
+/// Note: this does not fully compile on Solidity 0.8.8; only a concept.
+
+type UncheckedInt is int8;
+
+function add(UncheckedInt a, UncheckedInt b) pure returns(UncheckedInt) {
+    unchecked {
+        return UncheckedInt.wrap(UncheckedInt.unwrap(a) + UncheckedInt.unwrap(b));
+    }
+}
+
+using {add as +} for UncheckedInt;
+
+contract MockOperator {
+    UncheckedInt x;
+    function increment() external {
+        // This would not revert on overflow when x = 127
+        x = x + UncheckedInt.wrap(1);
+    }
+}
```

Comment:
```suggestion
function add(UncheckedInt a, UncheckedInt b) pure returns(UncheckedInt) {
    unchecked {
        return UncheckedInt.wrap(UncheckedInt.unwrap(a) + UncheckedInt.unwrap(b));
    }
}
function addInt(UncheckedInt a, uint b) pure returns(UncheckedInt) {
    unchecked {
        return UncheckedInt.wrap(UncheckedInt.unwrap(a) + b);
    }
}

using {add as +, addInt as +} for UncheckedInt;

contract MockOperator {
    UncheckedInt x;
    function increment() external {
        // This would not revert on overflow when x = 127
        x = x + 1;
    }
}
```